lib/core/src/Cardano/Wallet/Primitive/CoinSelection/MA/RoundRobin.hs

```diff
@@ -363,211 +642,256 @@ runSelectionStep lens s
+            changeForUnknownAssets = F.foldr
+                (NE.zipWith (<>) . makeChangeForUnknownAsset outputTokens)
+                (TokenMap.empty <$ outputTokens)
+                unknownAssets
+
+        let change :: NonEmpty TokenMap
+            change = NE.zipWith (<>) changeForKnownAssets changeForUnknownAssets
+
+        (bundles, remainder) <-
+            maybe (Left $ changeError excessCoin change) Right $
+                excessCoin `subtractCoin` requiredCost
+                >>=
+                runStateT
+                    (sequence (StateT . assignCoin minCoinValueFor <$> change))
+
+        let changeForCoins :: NonEmpty TokenBundle
+            changeForCoins = TokenBundle.fromCoin
+                <$> makeChangeForCoin outputCoins remainder
+
+        pure (NE.zipWith (<>) bundles changeForCoins)
   where
     totalInputValueInsufficient = error
         "makeChange: not (totalOutputValue <= totalInputValue)"
     totalOutputCoinValueIsZero = error
         "makeChange: not (totalOutputCoinValue > 0)"
 
+    changeError
+        :: Coin
+        -> NonEmpty TokenMap
+        -> UnableToConstructChangeError
+    changeError excessCoin change =
+        UnableToConstructChangeError
+            { missingCoins =
+                -- This conversion is safe because we know that the distance is
+                -- small-ish. If it wasn't, we would have have enough coins to
+                -- construct the change.
+                unsafeNaturalToCoin $ distance
+                    (coinToNatural excessCoin)
+                    (coinToNatural requiredCost + totalMinCoinValue)
+            }
+      where
+        totalMinCoinValue =
+            F.sum $ (coinToNatural . minCoinValueFor) <$> change
+
+    outputTokens :: NonEmpty TokenMap
+    outputTokens = view #tokens <$> outputBundles
+
+    outputCoins :: NonEmpty Coin
+    outputCoins = view #coin <$> outputBundles
+
     totalInputValue :: TokenBundle
-    totalInputValue = F.fold inputBundles
+    totalInputValue = TokenBundle.add
+        (F.fold inputBundles)
+        (maybe TokenBundle.empty TokenBundle.fromCoin mExtraCoinSource)
 
     totalOutputValue :: TokenBundle
     totalOutputValue = F.fold outputBundles
 
-    totalOutputCoinValue :: Coin
-    totalOutputCoinValue = TokenBundle.getCoin totalOutputValue
-
-    excess :: TokenBundle
-    excess =
-        -- The following subtraction is safe, as we have already checked that
-        -- the total input value is greater than the total output value:
-        totalInputValue `TokenBundle.unsafeSubtract` totalOutputValue
-
-    change :: NonEmpty TokenBundle
-    change
-        = TokenBundle.fromCoin <$> changeForCoin
-        & NE.zipWith (<>) (TokenBundle.fromTokenMap <$> changeForPaymentAssets)
-        -- Here we sort in ascending order of coin value so that surplus assets
-        -- and coin values are in increasing order in the resulting bundles:
-        & NE.sortWith TokenBundle.getCoin
```

Comment:
I'm not yet sure as to whether this property is maintained by the updated implementation, or whether we want to maintain it.

(Unfortunately, I did not have time to write a property test to check it.)

The original motivation was to avoid producing a UTxO distribution where either:
- large token quantities are bundled with tiny coin quantities; or
- large coin quantities are bundled with tiny token quantities.